src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,152 @@
+                forStatement.Incrementors.Count == 1)
+            {
+                var declaration = forStatement.Declaration;
+                if (declaration.Variables.Count == 1)
+                {
+                    var declarator = declaration.Variables[0];
+                    if (declarator.Initializer != null)
+                    {
+                        iterationVariable = declarator.Identifier;
+                        initializer = declarator.Initializer.Value;
+
+                        var binaryExpression = (BinaryExpressionSyntax)forStatement.Condition;
+
+                        // Look for:  i < expr.Length
+                        if (binaryExpression.Left is IdentifierNameSyntax identifierName &&
+                            identifierName.Identifier.ValueText == iterationVariable.ValueText &&
+                            binaryExpression.Right is MemberAccessExpressionSyntax)
+                        {
+                            memberAccess = (MemberAccessExpressionSyntax)binaryExpression.Right;
+
+                            var incrementor = forStatement.Incrementors[0];
+                            if (TryGetStepValue(iterationVariable, incrementor, out stepValue, cancellationToken))
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+
+            iterationVariable = default;
+            memberAccess = default;
+            initializer = default;
+            stepValue = default;
+            return false;
+        }
+
+        private static bool TryGetStepValue(
+            SyntaxToken iterationVariable, ExpressionSyntax incrementor,
+            out ExpressionSyntax stepValue, CancellationToken cancellationToken)
+        {
+            // support
+            //  x++
+            //  ++x
+            //  x += constant_1
+
+            ExpressionSyntax operand;
+            switch (incrementor.Kind())
+            {
+            case SyntaxKind.PostIncrementExpression:
+                operand = ((PostfixUnaryExpressionSyntax)incrementor).Operand;
+                stepValue = default;
+                break;
+
+            case SyntaxKind.PreIncrementExpression:
+                operand = ((PrefixUnaryExpressionSyntax)incrementor).Operand;
+                stepValue = default;
+                break;
+
+            case SyntaxKind.AddAssignmentExpression:
+                var assignment = (AssignmentExpressionSyntax)incrementor;
+                operand = assignment.Left;
+                stepValue = assignment.Right;
+                break;
+
+            default:
+                stepValue = null;
+                return false;
+            }
+
+            return operand is IdentifierNameSyntax identifierName &&
+                identifierName.Identifier.ValueText == iterationVariable.ValueText;
+        }
```

Comment:
few ideas:
1. could you use pattern matching instead of the casts? for assignment you'd have to use `when .IsKind()`, do you think that to be an improvement or not?
2. i don't like this imperative style with variables outside assigned from each case with a break everywhere, it seems like this would deserve maybe a local function or a method that could have a single return from each case with a tuple of (operand, stepValue)... in default it would return a null operand and the `operand is IdentifierNameSyntax ` would still work as expected

Reply:
i'm ok with it as it is.  If we had a match expression, i might consider using that.  But i'm ok with things as they are here.